File: android-app/src/main/java/com/medziku/plantwatering/data/parts/BatteryLevelSensorImpl.java
package com.medziku.plantwatering.data.parts;

import android.os.Parcel;
import android.os.Parcelable;

import com.medziku.plantwatering.station.parts.BatteryLevelSensor;


public class BatteryLevelSensorImpl extends BatteryLevelSensor implements Parcelable {

    public BatteryLevelSensorImpl(String name, float[] batteryLevels) {
        super(name, batteryLevels);
    }

    private BatteryLevelSensorImpl(Parcel in) {
        super(in.readString(), new float[]{});
        float[] floats = new float[in.readInt()];
        in.readFloatArray(floats);
        setBatteryLevels(floats);
    }

    public static final Creator<BatteryLevelSensorImpl> CREATOR = new Creator<BatteryLevelSensorImpl>() {
        @Override
        public BatteryLevelSensorImpl createFromParcel(Parcel in) {
            return new BatteryLevelSensorImpl(in);
        }

        @Override
        public BatteryLevelSensorImpl[] newArray(int size) {
            return new BatteryLevelSensorImpl[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(getName());
        dest.writeInt(getBatteryLevels().length);
        dest.writeFloatArray(getBatteryLevels());
    }
}

File: android-app/src/main/kotlin/com/medziku/plantwatering/data/parts/BatteryLevelSensorImpl.kt
package com.medziku.plantwatering.data.parts

import android.os.Parcel
import android.os.Parcelable

import com.medziku.plantwatering.station.parts.BatteryLevelSensor


class BatteryLevelSensorImpl : BatteryLevelSensor, Parcelable {

    constructor(name: String, batteryLevels: FloatArray) : super(name, batteryLevels)

    private constructor(`in`: Parcel) : super(`in`.readString(), floatArrayOf()) {
        val floats = FloatArray(`in`.readInt())
        `in`.readFloatArray(floats)
        batteryLevels = floats
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeInt(batteryLevels.size)
        dest.writeFloatArray(batteryLevels)
    }

    companion object {

        @JvmField
        val CREATOR: Parcelable.Creator<BatteryLevelSensorImpl> = object : Parcelable.Creator<BatteryLevelSensorImpl> {
            override fun createFromParcel(`in`: Parcel): BatteryLevelSensorImpl {
                return BatteryLevelSensorImpl(`in`)
            }

            override fun newArray(size: Int): Array<BatteryLevelSensorImpl?> {
                return arrayOfNulls(size)
            }
        }
    }
}